---
title: "RNA-Seq components (airway dataset)"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{iris}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r echo=F}
# To test
# devtools::load_all(".",export_all=F) ; rmarkdown::render("vignettes/airway.Rmd")

knitr::opts_chunk$set(fig.width=6, fig.height=4)
```

Let's look at the [airway](http://bioconductor.org/packages/release/data/experiment/html/airway.html) dataset as an example of a typical small-scale RNA-Seq experiment. In this experiment, four Airway Smooth Muscle (ASM) cell lines are treated with the asthma medication dexamethasone.

```{r setup, warning=F, message=F}
library(weitrix)
library(SummarizedExperiment)
library(EnsDb.Hsapiens.v86)
library(edgeR)
library(limma)
library(reshape2)
library(tidyverse)
library(airway)

BiocParallel::bpstart( DelayedArray::getAutoBPPARAM() )
set.seed(1234)
```

```{r}
data("airway")
airway
```


## Initial processing

Initial steps are the same as for a differential expression analysis.

```{r}
airway_dge <- 
    DGEList(assay(airway, "counts")) %>%
    calcNormFactors() 

airway_cpm <- cpm(airway_dge, log=TRUE, prior.count=1)

good <- rowMeans(airway_cpm) >= log2(1)
table(good)

design <- model.matrix(~ dex + cell, data=colData(airway))

airway_elist <- vooma(airway_cpm[good,], design=design, plot=TRUE)
```

Some of these steps are optional and have marginal value:

* `vooma` provides precision weights. One could instead choose a `prior.count` for `cpm` to produce a flat (or at least non-decreasing) mean-variance trend-line.

* Using a design matrix with `vooma` may not provide much benefit, and if you want an analysis completely blinded to the experimental design it might be better to omit this.

There are many options here, so adjust to your taste!


## Conversion to weitrix

```{r}
airway_weitrix <- as_weitrix(airway_elist)

# Include row and column information
colData(airway_weitrix) <- colData(airway)
rowData(airway_weitrix) <- 
    mcols(genes(EnsDb.Hsapiens.v86))[rownames(airway_weitrix),c("gene_name","gene_biotype")]

airway_weitrix
```


## Find components of variation

This will find various numbers of components, from 1 to 6. In each case, the components discovered have varimax rotation applied to their gene loadings to aid interpretability.

```{r message=F}
comp_seq <- weitrix_components_seq(airway_weitrix, p=6)
```

```{r message=F}
rand_weitrix <- weitrix_randomize(airway_weitrix)
rand_comp <- weitrix_components(rand_weitrix, p=1)

components_seq_screeplot(comp_seq, rand_comp)
```

## Examining components

```{r}
comp <- comp_seq[[4]]

comp$col[,-1] %>% melt(varnames=c("Run","component")) %>%
    left_join(as.data.frame(colData(airway)), by="Run") %>%
    ggplot(aes(y=cell, x=value, color=dex)) + 
    geom_vline(xintercept=0) + 
    geom_point(alpha=0.5, size=3) + 
    facet_grid(~ component) +
    labs(title="Sample scores for each component", x="Sample score", y="Cell line", color="Treatment")

comp$row[,-1] %>% melt(varnames=c("name","component")) %>%
    ggplot(aes(x=comp$row[name,"(Intercept)"], y=value)) + 
    geom_point(cex=0.1) + 
    facet_wrap(~ component) +
    labs(title="Gene loadings for each component vs average log2 RPM")
```

## Without varimax rotation, components may be harder to interpret

If varimax rotation isn't used, `weitrix_components` and `weitrix_components_seq` will produce a Principal Components Analysis, with components ordered from most to least variance explained.

Without varimax rotation the treatment effect is still mostly in the first component, but has also leaked a small amount into the other components.

```{r message=F}
comp_nonvarimax <- weitrix_components(airway_weitrix, p=4, use_varimax=FALSE)

comp_nonvarimax$col[,-1] %>% melt(varnames=c("Run","component")) %>%
    left_join(as.data.frame(colData(airway)), by="Run") %>%
    ggplot(aes(y=cell, x=value, color=dex)) + 
    geom_vline(xintercept=0) + 
    geom_point(alpha=0.5, size=3) + 
    facet_grid(~ component) +
    labs(title="Sample scores for each component, no varimax rotation", x="Sample score", y="Cell line", color="Treatment")
```


## `col` can potentially be used as a design matrix with limma

* If you are willing to place your trust in sparsity of effects, this may be a somewhat magical way of removing unwanted variation. 

* If you're not sure of the experimental design, for example the exact timing of a time series or how evenly a drug treatment was applied, the extracted component might actually be more accurate.

Note that this ignores uncertainty about the `col` matrix itself.

This may be useful for hypothesis generation -- finding some potentially interesting genes, while discounting noisy or lowly expressed genes -- but don't use it as proof of significance.

```{r}
airway_elist <- weitrix_elist(airway_weitrix)

fit <- 
    lmFit(airway_elist, comp$col) %>% 
    eBayes()

topTable(fit, "C1")

all_top <- topTable(fit, "C1", n=Inf, sort.by="none")
plotMD(fit, "C1", status=all_top$adj.P.Val <= 0.01)
```

You might also consider using my `topconfects` package. This will find the largest confident effect sizes, while still correcting for multiple testing.

```{r}
library(topconfects)
limma_confects(fit, "C1")
```




